docs/one_collect-design.md

```diff
@@ -0,0 +1,651 @@
+- CLR event processing
+- Managed/native boundary handling
+- .NET-specific data extraction
+
+##### Scripting Integration (`helpers::scripting`)
+
+The scripting engine integrates at the universal layer, allowing runtime customization of data processing. It hooks into the `ExportMachine` after all data has been aggregated:
+
+**Code Reference**: [`one_collect/src/helpers/exporting/scripting.rs`](one_collect/src/helpers/exporting/scripting.rs)
+
+```rust
+use one_collect::helpers::exporting::*;
+
+// Scripting hooks into the universal layer
+let mut exporter = UniversalExporter::new(settings);
+
+// Add script that processes aggregated data
+exporter.add_parsed_hook(|context: &mut UniversalParsedContext| {
+    let machine = context.machine_mut();
+    
+    // Script has access to all aggregated data
+    // and can perform custom analysis/filtering
+    machine.apply_custom_filter(|record| {
+        // Custom filtering logic
+        true
+    })?;
+    
+    machine.add_custom_metric("my_metric", calculate_metric(&machine))?;
+    Ok(())
+});
+
+// Result is an ExportMachine with all data aggregated
+// and custom processing applied
+let final_machine = exporter.process_with_scripts()?;
+```
+
+#### Utility Modules
+
+##### Interning (`intern` module)
+Memory-efficient string and data management through deduplication:
+
+**Code Reference**: [`one_collect/src/intern.rs`](one_collect/src/intern.rs)
+
+###### String Interning
+```rust
+use one_collect::intern::InternedStrings;
+
+// Create an interned strings container
+let mut strings = InternedStrings::new();
+
+// Store strings with deduplication
+let id1 = strings.intern("kernel32.dll");
+let id2 = strings.intern("kernel32.dll"); // Same string, same ID
+let id3 = strings.intern("ntdll.dll");    // Different string, different ID
+
+assert_eq!(id1, id2); // Same ID for identical strings
+
+// Retrieve strings by ID
+let name = strings.get(id1).unwrap();
+println!("Module: {}", name); // "kernel32.dll"
+```
+
+###### Callstack Interning
+```rust
+use one_collect::intern::InternedCallstacks;
+
+// Create a callstack interning container
+let mut callstacks = InternedCallstacks::new();
+
+// Store callstack with automatic deduplication
+let addresses = vec![0x7FF123456789, 0x7FF987654321, 0x7FF111222333];
+let callstack_id = callstacks.intern(addresses.clone());
+
+// Same callstack gets same ID
+let duplicate_id = callstacks.intern(addresses);
+assert_eq!(callstack_id, duplicate_id);
+
+// Access the stored callstack
+if let Some(stack) = callstacks.get(callstack_id) {
+    for addr in stack {
+        println!("Frame: 0x{:x}", addr);
+    }
+}
+```
+
+## Data Flow Architecture
+
+### Event Processing Pipeline
+
+1. **Event Source**: Platform-specific event collection (perf events, ETW)
+2. **Event Registration**: Events registered with format specifications
+3. **Data Ingestion**: Raw event data received from OS
+4. **Format Parsing**: Data interpreted according to event format
```

Comment:
Mark this as optional, since it doesn't always happen - only when necessary based on the pipeline.